osquery/tables/system/windows/users.cpp

```diff
@@ -24,81 +24,148 @@
     wchar_t accntName[UNLEN] = {0};
     wchar_t domName[DNLEN] = {0};
     unsigned long accntNameLen = UNLEN;
     unsigned long domNameLen = DNLEN;
     SID_NAME_USE eUse;
-
-    /// MSDN guaruntees the string values returned is null terminated.
     ret = LookupAccountSidW(
         nullptr, sid, accntName, &accntNameLen, domName, &domNameLen, &eUse);
-    if (ret != 0) {
-      r["username"] = wstringToString(accntName);
-    } else {
-      TLOG << "Lookup Account by SID failed with: " << GetLastError();
-    }
+    r["username"] = ret != 0 ? wstringToString(accntName) : "";
+  }
+  results.push_back(r);
+}
 
-    /// USER_INFO_3 contains detailed info, like the uid
-    unsigned long dwDetailedUserLevel = 3;
-    LPUSER_INFO_3 pUserInfoBuffer = nullptr;
-
-    NET_API_STATUS nStatus;
-    nStatus = NetUserGetInfo(nullptr,
-                             accntName,
-                             dwDetailedUserLevel,
-                             reinterpret_cast<LPBYTE*>(&pUserInfoBuffer));
-
-    if (nStatus == NERR_Success) {
-      r["uid"] = INTEGER(pUserInfoBuffer->usri3_user_id);
-      r["uid_signed"] = INTEGER(pUserInfoBuffer->usri3_user_id);
-      r["gid"] = INTEGER(pUserInfoBuffer->usri3_primary_group_id);
-      r["gid_signed"] = INTEGER(pUserInfoBuffer->usri3_primary_group_id);
-      r["description"] =
-          SQL_TEXT(wstringToString(pUserInfoBuffer->usri3_comment));
-    } else {
-      /// If NetUserGetInfo fails parse the uid manually
-      auto sidTokens = osquery::split(sidString, "-");
-      if (sidTokens.size() == 8) {
-        r["uid"] = INTEGER(sidTokens.back());
-        r["uid_signed"] = INTEGER(sidTokens.back());
+void processWinLocalAccounts(std::set<std::string>& processedSids,
+                             QueryData& results) {
+  unsigned long dwUserInfoLevel = 3;
+  unsigned long dwNumUsersRead = 0;
+  unsigned long dwTotalUsers = 0;
+  unsigned long resumeHandle = 0;
+  unsigned long ret = 0;
+  LPBYTE userBuffer = nullptr;
+  do {
+    ret = NetUserEnum(nullptr,
+                      dwUserInfoLevel,
+                      0,
+                      &userBuffer,
+                      MAX_PREFERRED_LENGTH,
+                      &dwNumUsersRead,
+                      &dwTotalUsers,
+                      &resumeHandle);
+
+    if ((ret == NERR_Success || ret == ERROR_MORE_DATA) &&
+        userBuffer != nullptr) {
+      auto iterBuff = LPUSER_INFO_3(userBuffer);
+      for (size_t i = 0; i < dwNumUsersRead; i++) {
+        // User level 4 contains the SID value
+        unsigned long dwDetailedUserInfoLevel = 4;
+        LPBYTE userLvl4Buff = nullptr;
+        ret = NetUserGetInfo(nullptr,
+                             iterBuff->usri3_name,
+                             dwDetailedUserInfoLevel,
+                             &userLvl4Buff);
+
+        if (ret != NERR_Success || userLvl4Buff == nullptr) {
+          NetApiBufferFree(userLvl4Buff);
```

Comment:
I would add another check to make sure `userLbl4Buff != nullptr` before trying to free it.